### PR TITLE
response return value runtime check

### PR DESCRIPTION
**Pull request by benitogf**
https://github.com/aio-libs/aiohttp/issues/3311

### CODE REVIEW

**Comment by asvetlov** [comment]
Also, a test is required to never miss the functionality after code refactoring.

**Comment by benitogf** [comment]
> The check should be performed in debug mode only for sake of performance.

I don't follow this part, how is checking if we are in debug mode different from checking the resp type?

w/o debug check:

```
                    if isinstance(resp, StreamResponse):
                        await resp.prepare(request)
                        await resp.write_eof()
                    else:
                        raise RuntimeError('Possibly missing return ' +
                                        ' statement on request handler')
```

with debug check:

```
                if self._loop.get_debug():
                    if isinstance(resp, StreamResponse):
                        await resp.prepare(request)
                        await resp.write_eof()
                    else:
                        raise RuntimeError('Possibly missing return ' +
                                        ' statement on request handler')
                else:
                    await resp.prepare(request)
                    await resp.write_eof()
```

**Comment by asvetlov** [comment]
```
                if self.debug:
                    if inot sinstance(resp, StreamResponse):
                        raise RuntimeError('Web-handler should return a response instance, got {!r}'.format(resp))
                await resp.prepare(request)
                await resp.write_eof()
```

File: aiohttp/web_protocol.py [hunk 1]
```diff
@@ -438,7 +445,7 @@ def _process_keepalive(self):
                 self.log_exception('Unhandled exception', exc_info=exc)
                 self.force_close()
             finally:
-                if self.transport is None:
+                if self.transport is None and resp is not None:
```
[the remark this answers is not in the document]
Comment (benitogf) [reply aiohttp/web_protocol.py:449]: this check prevents the `log.debug` last value being overwritten by: https://github.com/aio-libs/aiohttp/pull/3321/files/8d585c826f9fb6fc82a5291f5973caf136aaabee#diff-8ac506a3012476288a758f156ca7d7b2R449 and fail the assert: https://github.com/aio-libs/aiohttp/pull/3321/files/8d585c826f9fb6fc82a5291f5973caf136aaabee#diff-98519d3deafe019f57304a557d20531cR542

maybe there's a way to test it w/o relying on the last `log.debug` value?

[the remark this answers is not in the document]
Comment (asvetlov) [reply aiohttp/web_protocol.py:449]: Got you.
Let's drop `self.log_debug("Possibly missing return statement on request handler")` log but log `RuntimeError` with it's message.

`'Web-handler should return a response instance, got {!r}'` is a good enough hint to the missing return statement, isn't it?

File: aiohttp/web_protocol.py [hunk 2]
```diff
@@ -389,6 +390,12 @@ def _process_keepalive(self):
                             "please raise the exception instead",
                             DeprecationWarning)
 
+                if self.debug:
+                    if not isinstance(resp, StreamResponse):
+                        self.log_debug("Possibly missing return "
+                                       "statement on request handler")
+                        raise RuntimeError('Web-handler should \
```
Comment (asvetlov) [inline aiohttp/web_protocol.py]: Please replace `\` with implicit strings concatenation as a line above.

File: aiohttp/web_protocol.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (benitogf) [reply aiohttp/web_protocol.py:449]: nope, that runtimeError will be bypassed by this one: https://github.com/aio-libs/aiohttp/pull/3321/files/8d585c826f9fb6fc82a5291f5973caf136aaabee#diff-8ac506a3012476288a758f156ca7d7b2R439 another reason why I used `log.debug`

[the remark this answers is not in the document]
Comment (asvetlov) [reply aiohttp/web_protocol.py:449]: Ok. Agree

**Comment by asvetlov** [comment]
thanks!

**Pull request merged**